src/System.Management.Automation/engine/hostifaces/PSTask.cs

```diff
@@ -763,18 +802,28 @@ public bool Add(PSTaskChildJob childJob)
         /// </summary>
         public void StopAll()
         {
+            _stopping = true;
+
             // Accept no more input
             Close();
             _stopAll.Set();
 
             // Stop all running tasks
+            PSTaskBase[] tasksToStop;
             lock (_syncObject)
             {
-                foreach (var task in _taskPool.Values)
-                {
-                    task.Dispose();
-                }
+                tasksToStop = new PSTaskBase[_taskPool.Values.Count];
+                _taskPool.Values.CopyTo(tasksToStop, 0);
```

Comment:
I'm curious why this copy is needed. It's a reference copy, right? So both `tasksToStop` and `_taskPool` will pointing to the same set of `PSTaskBase` objects.

Reply:
Yes, but _taskPool dictionary is modified when a task is disposed. Also the task must be disposed outside the _syncObject lock to avoid a deadlock. To solve these two problems I create a temporary local copy and use that to dispose the final list of runspaces.

Reply:
Got it. So is there a possibility that new task get added to `_taskPool` after the copy is done? In that case, the new task will not be disposed, but I guess it will be handled elsewhere?

Reply:
I don't like to use the word 'impossible' with multi-threading, but it should not be possible to add anymore tasks since the pool is closed and adding tasks is done on a single thread.